### PR TITLE
cli(add): support for webpack config file

### LINKED ISSUE #279: Specify config file with webpack-cli add
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
We may have webpack config files with different base names. However, commands except for `init`, such as `add`, always takes `webpack.config.js` as the processed file. It will be better if we can specify the path to config file for these commands.    
Affected commands: `add`(current version), `update`, `remove`(future)

**If this is a feature request, what is motivation or use case for changing the behavior?**
Take command `add` as an example:
```
$ webpack-cli add <webpack_config_file>
```
To add components to the specified config file.
This usage is similar to existing command `migrate`.

**Please mention other relevant information such as the browser version, Node.js version, Operating System and programming language.**

### PR DESCRIPTION
**Pull request by dhruvdutt**
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
No

**If relevant, did you update the documentation?**
Yes

**Summary**
Add support to specify webpack config file with args, default to webpack.config.js

- Specifying config file: `webpack-cli add webpack.dev.js`

image

- Default config file: `webpack-cli add`

image

- Non-existential file: `webpack-cli add aaahha.js`

image

Similar usage for `update`, `remove`.

**Does this PR introduce a breaking change?**
No

Fixes #279

### CODE REVIEW

**Comment by dhruvdutt** [comment]
@ev1stensberg Already rebased.

image

File: lib/commands/migrate.js [hunk 1]
```diff
@@ -14,23 +14,18 @@ const WebpackOptionsValidationError = require("webpack")
...
-* @returns {Promise} Runs the migration using a promise that will throw any errors during each transform
-* or output if the user decides to abort the migration
-*/
-
-module.exports = function migrate(
-	currentConfigPath,
-	outputConfigPath,
-	options
-) {
+module.exports = function migrate(configPath, options) {
```
Comment (ematipico) [inline lib/commands/migrate.js]: I don't get this change. Why a parameter has been removed?

File: lib/index.js [hunk 2]
```diff
@@ -7,39 +7,36 @@ const path = require("path");
...
+ * @param	{Object}	args - arguments from the CLI
+ * @returns	{Function}	invokes the module with the supplied command
  *
  */
 
 module.exports = function initialize(command, args) {
 	const popArgs = args ? args.slice(2).pop() : null;
+	const cmdArgs = args.slice(2).length === 1 ? [] : [popArgs];
+	// default config filename
+	const configFile = cmdArgs.length ? cmdArgs[0] : "webpack.config.js";
```
Comment (ematipico) [inline lib/index.js]: `"webpack.config.js"` should be inside a constant
```js
const DEFAULT_WEBPACK_CONFIG_FILENAME = "webpack.config.js";
```
A bit long but it gives clear idea lol

File: lib/commands/migrate.js (hunk 1 above)
Comment (dhruvdutt) [reply lib/commands/migrate.js]: https://github.com/webpack/webpack-cli/blob/b6a30ad94a100440b040b3533d90960582bc42b0/lib/index.js#L31

We were using the same variable. It was redundant to have 2 params for a value coming from a single source.

File: lib/index.js (hunk 2 above)
Comment (dhruvdutt) [reply lib/index.js]: Added the variable.

File: lib/commands/migrate.js (hunk 1 above)
Comment (ematipico) [reply lib/commands/migrate.js]: So maybe the bug is actually there when we call migrate.js, not here. I don't think we should change the signature of the function, we are basically changing the semantic and functionality of the function, I think.

Comment (dhruvdutt) [reply lib/commands/migrate.js]: I agree but AFAIK we aren't using migrate module at any other places in the code. This should fix the new params signature.

**Comment by ematipico** [comment]
We can now rebase this PR from master. Also, could you add a test case?

File: lib/commands/add.js [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
  *
  */
 
-module.exports = function add() {
-	return modifyHelper("add", defaultGenerator);
+const DEFAULT_WEBPACK_CONFIG_FILENAME = "webpack.config.js";
+
+module.exports = function add(...args) {
+	const filePaths = args.slice(3);
+	let configFile = DEFAULT_WEBPACK_CONFIG_FILENAME;
```
Comment (ev1stensberg) [inline lib/commands/add.js]: No need for outer scope constant here

File: lib/generators/init-generator.js [hunk 4]
```diff
@@ -50,7 +50,7 @@ module.exports = class InitGenerator extends Generator {
 		process.stdout.write(
 			"\n" +
 				logSymbols.info +
-				chalk.blue(" INFO ") +
+				chalk.blue(" info ") +
```
Comment (ev1stensberg) [inline lib/generators/init-generator.js]: no need

File: lib/generators/init-generator.js [hunk 5]
```diff
@@ -59,7 +59,7 @@ module.exports = class InitGenerator extends Generator {
 		);
 		process.stdout.write(
 			logSymbols.info +
-				chalk.blue(" INFO ") +
+				chalk.blue(" info ") +
```
Comment (ev1stensberg) [inline lib/generators/init-generator.js]: no need

File: lib/generators/loader-generator.js [hunk 6]
```diff
@@ -49,9 +49,7 @@ const LoaderGenerator = webpackGenerator(
 		"examples/simple/src/lazy-module.js.tpl",
 		"examples/simple/src/static-esm-module.js.tpl"
 	],
-	[
-		"src/_index.js.tpl"
-	],
+	["src/_index.js.tpl"],
```
Comment (ev1stensberg) [inline lib/generators/loader-generator.js]: no need

File: lib/generators/plugin-generator.js [hunk 7]
```diff
@@ -30,10 +30,7 @@ const PluginGenerator = webpackGenerator(
 		"examples/simple/src/lazy-module.js.tpl",
 		"examples/simple/src/static-esm-module.js.tpl"
 	],
-	[
-		"src/_index.js.tpl",
-		"examples/simple/_webpack.config.js.tpl"
-	],
+	["src/_index.js.tpl", "examples/simple/_webpack.config.js.tpl"],
```
Comment (ev1stensberg) [inline lib/generators/plugin-generator.js]: no need

File: lib/generators/loader-generator.js (hunk 6 above)
Comment (dhruvdutt) [reply lib/generators/loader-generator.js]: I think `format` script / prettier the auto-formatting. Should I still restore to older formatting?

File: lib/generators/init-generator.js (hunk 4 above)
Comment (dhruvdutt) [reply lib/generators/init-generator.js]: I like how yarn shows the message type in lowercase. Again, we can do either way. We just need to follow same consistency for all types of messages.
 
image

File: lib/commands/add.js (hunk 3 above)
Comment (dhruvdutt) [reply lib/commands/add.js]: @ematipico [suggested](https://github.com/webpack/webpack-cli/pull/352#discussion_r175936671) to keep a constant. I'm convinced either way. :sweat_smile:

File: lib/generators/loader-generator.js (hunk 6 above)
Comment (ev1stensberg) [reply lib/generators/loader-generator.js]: y discard this

File: lib/generators/init-generator.js (hunk 4 above)
Comment (ev1stensberg) [reply lib/generators/init-generator.js]: could you src one with the old info text too?

File: lib/commands/add.js (hunk 3 above)
Comment (ev1stensberg) [reply lib/commands/add.js]: store the constant within the scope of the function

Comment (dhruvdutt) [reply lib/commands/add.js]: Fixed.

File: lib/generators/init-generator.js (hunk 5 above)
Comment (dhruvdutt) [reply lib/generators/init-generator.js]: Reverted.

File: lib/generators/loader-generator.js (hunk 6 above)
Comment (dhruvdutt) [reply lib/generators/loader-generator.js]: I think CI would anyway run `npm run format` or `yarn format` and write the files so it's better to have it same in the source code to have consistency. :)

**Comment by dhruvdutt** [comment]
Let's merge this. :)

**Comment by ev1stensberg** [comment]
This is labeled `chill` @dhruvdutt :)

**Comment by ev1stensberg** [comment]
You'll need to fix conflicts here.

**Comment by dhruvdutt** [comment]
Seems like lots of things have messed up as this was initially prepared against master. Coming back to this tomorrow. :sweat_smile:

**Comment by ev1stensberg** [comment]
What are you working on today?

**Comment by dhruvdutt** [comment]
@ev1stensberg, Working on add parser majorly. Can you please check Slack?

**Comment by ev1stensberg** [comment]
I think you'll need to reopen this against the new `next`. This PR is way to flaky to be merged in reasonable time, a lot of merge conflicts and your implementation makes the `init` feature throw when it shouldn't throw.

**Pull request closed without merging**